**Necessary Imports**

In [45]:
import pandas as pd 
import numpy as np 
import re 
import matplotlib.pyplot as plt

**Loading the Necessary Files**

In [46]:
df = pd.read_csv("../Data/car.csv")
df.head()

,Name,Price,Used For,Transmisson,Colour,Make Year,Mileage,Engine (CC),Fuel,Kilometer Run,Waranty,Types
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,"रू. 24,75,000रू. 25,00,000",Private Use,Manual2WD,Brown,2015,14,1400,Petrol,42000,NaN,NaN
1,Excellent car on sale (Hyundai),"रू. 7,50,000",NaN,Auto2WD,Light blue,2005,11,1399,Petrol,87412,NaN,NaN
2,TATA 407 Container (Tata),"रू. 7,00,000",NaN,Manual - 2WD,White,2013,NaN,2956,Diesel,60000,NaN,NaN
3,4x4 swaraj Mazda (Mahindra),"रू. 6,00,000",NaN,Manual - 4WD,NaN,2017,NaN,NaN,Diesel,NaN,NaN,NaN
4,i20 Active good for used few time (Hyundai),रू. 375,NaN,Auto - 2WD,white,2019,17,1200,Petrol,2400,NaN,NaN


In [47]:
print(len(df))

848


**Changing the columns into Snake Case**

In [48]:
df.columns = (
    df.columns               
      .str.lower()                  # convert to lowercase
      .str.replace('[^a-zA-Z0-9 ]', '', regex=True)   # remove special chars ()&./
      .str.replace(' +', '_', regex=True)  # convert spaces to underscores
)
df.head()

,name,price,used_for,transmisson,colour,make_year,mileage,engine_cc,fuel,kilometer_run,waranty,types
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,"रू. 24,75,000रू. 25,00,000",Private Use,Manual2WD,Brown,2015,14,1400,Petrol,42000,NaN,NaN
1,Excellent car on sale (Hyundai),"रू. 7,50,000",NaN,Auto2WD,Light blue,2005,11,1399,Petrol,87412,NaN,NaN
2,TATA 407 Container (Tata),"रू. 7,00,000",NaN,Manual - 2WD,White,2013,NaN,2956,Diesel,60000,NaN,NaN
3,4x4 swaraj Mazda (Mahindra),"रू. 6,00,000",NaN,Manual - 4WD,NaN,2017,NaN,NaN,Diesel,NaN,NaN,NaN
4,i20 Active good for used few time (Hyundai),रू. 375,NaN,Auto - 2WD,white,2019,17,1200,Petrol,2400,NaN,NaN


**Dropping Duplicate**

In [49]:
df.drop_duplicates(subset=["name"], keep="first")


,name,price,used_for,transmisson,colour,make_year,mileage,engine_cc,fuel,kilometer_run,waranty,types
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,"रू. 24,75,000रू. 25,00,000",Private Use,Manual2WD,Brown,2015,14,1400,Petrol,42000,NaN,NaN
1,Excellent car on sale (Hyundai),"रू. 7,50,000",NaN,Auto2WD,Light blue,2005,11,1399,Petrol,87412,NaN,NaN
2,TATA 407 Container (Tata),"रू. 7,00,000",NaN,Manual - 2WD,White,2013,NaN,2956,Diesel,60000,NaN,NaN
3,4x4 swaraj Mazda (Mahindra),"रू. 6,00,000",NaN,Manual - 4WD,NaN,2017,NaN,NaN,Diesel,NaN,NaN,NaN
4,i20 Active good for used few time (Hyundai),रू. 375,NaN,Auto - 2WD,white,2019,17,1200,Petrol,2400,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
843,Hyundai grand i10 (Hyundai),"रू. 24,50,000",2 years,Manual - 2WD,silver,2020,2025,1200,Petrol,26500,NaN,NaN
844,Well maintained car. Everything is fine. Singl...,"रू. 19,25,000",6 years,Manual2WD,Navy blue,2015,15,1200,Petrol,60000,No,NaN
845,VW Polo 2014 Highline 1200 (Renault),"रू. 19,50,000",7,Manual2WD,Carbon Steel ( Dark Grey),2014,12 to 14,1200,Petrol,60000,NaN,NaN
846,Hyundai i10 magna 2014 (Hyundai),"रू. 17,50,000",NaN,Manual - 2WD,silver,2014,NaN,1200,Petrol,93000,NaN,NaN


**Changing the KW to cc**

In [50]:
def convert_kw_to_cc(value):
    if pd.isna(value):
        return value
    
    # Convert to string for processing
    value_str = str(value).strip().lower()
    
    # Check if 'kw' is in the value
    if 'kw' in value_str:
        # Extract numeric part using regex
        match = re.search(r'(\d+\.?\d*)', value_str)
        if match:
            numeric_value = float(match.group(1))
            # Multiply by 24.25 and return
            return numeric_value * 24.25
    
    # Return original value if no 'kw' found
    return value

df['engine_cc']= df['engine_cc'].apply(convert_kw_to_cc)


**Cleaning Numeric Columns**

In [51]:

def clean_numeric_columns(df, columns):
    for col in columns:
        df[col] = df[col].astype(str).str.extract(r'(\d+(?:\.\d+)?)')
        df[col] = pd.to_numeric(df[col], errors='coerce')
    return df

df = clean_numeric_columns(df,['make_year','mileage','engine_cc','kilometer_run'])
df.head()

,name,price,used_for,transmisson,colour,make_year,mileage,engine_cc,fuel,kilometer_run,waranty,types
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,"रू. 24,75,000रू. 25,00,000",Private Use,Manual2WD,Brown,2015,14.0,1400.0,Petrol,42000.0,NaN,NaN
1,Excellent car on sale (Hyundai),"रू. 7,50,000",NaN,Auto2WD,Light blue,2005,11.0,1399.0,Petrol,87412.0,NaN,NaN
2,TATA 407 Container (Tata),"रू. 7,00,000",NaN,Manual - 2WD,White,2013,NaN,2956.0,Diesel,60000.0,NaN,NaN
3,4x4 swaraj Mazda (Mahindra),"रू. 6,00,000",NaN,Manual - 4WD,NaN,2017,NaN,NaN,Diesel,NaN,NaN,NaN
4,i20 Active good for used few time (Hyundai),रू. 375,NaN,Auto - 2WD,white,2019,17.0,1200.0,Petrol,2400.0,NaN,NaN


**Since, ['engine_cc','make_year','price','mileage','kilometer_run'] are the critical columns thus, we are imputing to channge them **

In [52]:
def sixty_percent_or_more_columns_missing(df,critical_columns):
    missing_count = df[critical_columns].isna().sum(axis=1)
    to_drop = missing_count >= 3
    print(f"Length of outliers:{len(df.loc[to_drop, critical_columns])}\n\n")


In [53]:
sixty_percent_or_more_columns_missing(df,['engine_cc', 'make_year', 'price','mileage','kilometer_run'])

Length of outliers:61




**Using capped_mputation of kilometers_run on the basis of make_year**

In [54]:
df['make_year'] = pd.to_numeric(df['make_year'], errors='coerce')
df['kilometer_run'] = pd.to_numeric(df['kilometer_run'], errors='coerce')
# Calculate actual km/year from existing data
df['km_per_year'] = df['kilometer_run'] / (2024 - df['make_year'])

# Get realistic median km/year from your dataset
cap_value = 66000 + 1.5 * 34000  # Use your actual rate but cap at Q3 + 1.5*IQR (standard outlier threshold)
actual_km_per_year = df['km_per_year'].median()  # Likely 5,000-8,000
print(f"Actual km/year in dataset: {actual_km_per_year}")

# Now impute using YOUR data's pattern
missing_mask = df['kilometer_run'].isna()
df.loc[missing_mask, 'kilometer_run'] = np.minimum(
    (2024 - df.loc[missing_mask, 'make_year']) * 4788,
    cap_value
)
df.drop('km_per_year',axis=1,inplace=True)


Actual km/year in dataset: 4777.777777777777


In [55]:
sixty_percent_or_more_columns_missing(df,['engine_cc', 'make_year', 'price','mileage','kilometer_run'])

Length of outliers:0




**Imputing Mileage on the basis of engine_cc**

In [56]:
df['engine_cc'] = pd.to_numeric(df['engine_cc'], errors='coerce')
# Group by fuel type and engine size categories
df['engine_category'] = pd.cut(df['engine_cc'], 
                                bins=[0, 1200, 1600, 2000, 5000],
                                labels=['Small', 'Medium', 'Large', 'Very Large'])

# Impute with group median
df['mileage'] = df.groupby(['fuel', 'engine_category'])['mileage'].transform(
    lambda x: x.fillna(x.median())
)

df.drop('engine_category', axis=1, inplace=True)

# Check result
print(df['mileage'].describe())


count      736.000000
mean       203.722201
std       3186.759449
min          5.000000
25%         12.000000
50%         14.000000
75%         15.000000
max      69000.000000
Name: mileage, dtype: float64


C:\Users\LEGION\AppData\Local\Temp\ipykernel_17468\3376968997.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df['mileage'] = df.groupby(['fuel', 'engine_category'])['mileage'].transform(


**Dropping 'colour' because it has no semantic value, 'types' because it is complektely null, 'waranty' because it is mostly null and 'used_for' because it is very variable **

In [57]:
df.drop(['used_for','colour','waranty','types'],axis=1,inplace=True)

**Removing Currency Symbol**

In [58]:
df['price'] = df['price'].str.replace("रू.","",regex=False).str.replace(",","",regex=False).str.strip() #removes the rupee symbol and commas in the price. 
df.loc[0,'price']=2475000
df.head()

,name,price,transmisson,make_year,mileage,engine_cc,fuel,kilometer_run
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,2475000,Manual2WD,2015,14.0,1400.0,Petrol,42000.0
1,Excellent car on sale (Hyundai),750000,Auto2WD,2005,11.0,1399.0,Petrol,87412.0
2,TATA 407 Container (Tata),700000,Manual - 2WD,2013,12.0,2956.0,Diesel,60000.0
3,4x4 swaraj Mazda (Mahindra),600000,Manual - 4WD,2017,NaN,NaN,Diesel,33516.0
4,i20 Active good for used few time (Hyundai),375,Auto - 2WD,2019,17.0,1200.0,Petrol,2400.0


There are some listings with name in Nepali but they are kept as it is because it is very difficult to extend semantic meaning from them. 

**Changing the transmission column into transmission_type and drive**

In [59]:
def split_transmisson(value):
    
    #first, let us check if the value is none. 
    if pd.isna(value):
        return (None,None)

    #Lets seperate by space
    value = value.strip()

    #If - is in the value then seperate by that value
    if '-' in value:
        parts = [p.strip() for p in value.split('-')]
        if len(parts) == 2:
            transmission_type = parts[0]
            drive = parts[1]
            return (transmission_type, drive)

    #Finally if the - doesn't exist we seperate alphabetic and numeric/WD parts 
    match = re.match(r"([A-Za-z]+)(.*)", value)
    if match:
        transmission_type = match.group(1)
        drive = match.group(2)
        return(transmission_type,drive)

    return(value,None)  

df[['transmisson_type','drive']] = df['transmisson'].apply(lambda x: pd.Series(split_transmisson(x)))
df.drop('transmisson', axis=1, inplace=True)
df.head() 


,name,price,make_year,mileage,engine_cc,fuel,kilometer_run,transmisson_type,drive
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,2475000,2015,14.0,1400.0,Petrol,42000.0,Manual,2WD
1,Excellent car on sale (Hyundai),750000,2005,11.0,1399.0,Petrol,87412.0,Auto,2WD
2,TATA 407 Container (Tata),700000,2013,12.0,2956.0,Diesel,60000.0,Manual,2WD
3,4x4 swaraj Mazda (Mahindra),600000,2017,NaN,NaN,Diesel,33516.0,Manual,4WD
4,i20 Active good for used few time (Hyundai),375,2019,17.0,1200.0,Petrol,2400.0,Auto,2WD


**Discovery of Outliers in the Numeric Data**

In [60]:
def find_quartiles_and_outliers(columns):
    for column in columns:
        print(f"\n\n For the {column}")
        df[column] = pd.to_numeric(df[column],errors='coerce')
        Q1 = df[column].quantile(0.25)
        Q2 = df[column].mean()
        Q3 = df[column].quantile(0.75)
        IQR = Q3 - Q1
        print(f"Q1:{Q1} ,Q2:{Q2} ,Q3:{Q3} , IQR:{IQR}") 
        outliers = df[(df[column] < (Q1 - 1.5* IQR)) | (df[column] > (Q3 + 1.5 * IQR))]
        print(f"Number of Outliers: {len(outliers)}")

find_quartiles_and_outliers(['kilometer_run','engine_cc','price','mileage','make_year'])



 For the kilometer_run
Q1:32891.0 ,Q2:57927.2712264151 ,Q3:67274.0 , IQR:34383.0
Number of Outliers: 22


 For the engine_cc
Q1:1197.0 ,Q2:1520.64689608637 ,Q3:1600.0 , IQR:403.0
Number of Outliers: 112


 For the price
Q1:1175000.0 ,Q2:2334365.9572446556 ,Q3:2650000.0 , IQR:1475000.0
Number of Outliers: 68


 For the mileage
Q1:12.0 ,Q2:203.7222010869565 ,Q3:15.0 , IQR:3.0
Number of Outliers: 39


 For the make_year
Q1:2010.0 ,Q2:1999.9870283018868 ,Q3:2017.0 , IQR:7.0
Number of Outliers: 22


Special for Kilometers Run

In [61]:
def find_quartiles_and_outliers_print(columns):
    for column in columns:
        print(f"\n\n For the {column}")
        df[column] = pd.to_numeric(df[column],errors='coerce')
        Q1 = df[column].quantile(0.25)
        Q2 = df[column].mean()
        Q3 = df[column].quantile(0.75)
        IQR = Q3 - Q1
        print(f"Q1:{Q1} ,Q2:{Q2} ,Q3:{Q3} , IQR:{IQR}") 
        outliers = df[(df[column] < (Q1 - 1.5* IQR)) | (df[column] > (Q3 + 1.5 * IQR))]
        print(outliers[column])
        print(f"Number of Outliers: {len(outliers)}")



In [62]:
find_quartiles_and_outliers_print(['kilometer_run'])



 For the kilometer_run
Q1:32891.0 ,Q2:57927.2712264151 ,Q3:67274.0 , IQR:34383.0
37      750840.0
175     610000.0
222     130000.0
267     790000.0
271     155000.0
311    1700000.0
325     146000.0
427     485780.0
441     320000.0
476     127130.0
486     125000.0
491     560000.0
541     160000.0
557     143000.0
586     125000.0
588     140000.0
618     750000.0
656     150000.0
701    -172368.0
784     149000.0
811    1350000.0
820     180000.0
Name: kilometer_run, dtype: float64
Number of Outliers: 22


There is one negative value, making it positive and Removing the values more than 180000, since they are in millions.

In [63]:
df.loc[701,'kilometer_run']=172368

In [64]:
upper_bound=180000
df['kilometer_run'] = df['kilometer_run'].clip(upper=upper_bound)
find_quartiles_and_outliers_print(['kilometer_run'])



 For the kilometer_run
Q1:33000.0 ,Q2:51616.08726415094 ,Q3:68000.0 , IQR:35000.0
37     180000.0
175    180000.0
222    130000.0
267    180000.0
271    155000.0
311    180000.0
325    146000.0
427    180000.0
441    180000.0
476    127130.0
486    125000.0
491    180000.0
541    160000.0
557    143000.0
586    125000.0
588    140000.0
618    180000.0
656    150000.0
701    172368.0
784    149000.0
811    180000.0
820    180000.0
Name: kilometer_run, dtype: float64
Number of Outliers: 22


**Special for Make Year**

In [65]:
find_quartiles_and_outliers_print(['make_year'])



 For the make_year
Q1:2010.0 ,Q2:1999.9870283018868 ,Q3:2017.0 , IQR:7.0
42     1997
127    1975
255    1998
266    1997
275    1997
322    1998
325    1998
465    1975
472    1999
488       0
503    2069
565       0
575     207
603    1992
612    1999
661    1998
701    2060
711     202
757      13
805     207
825    1993
827    1999
Name: make_year, dtype: int64
Number of Outliers: 22


Dropping 0 values, Manually changing 2069 and 2060 to AD, 13 to 2013 and 207 and 202 to 2007 and 2002

In [66]:
df = df.drop([488,565])
find_quartiles_and_outliers_print(['make_year'])



 For the make_year
Q1:2010.0 ,Q2:2004.7151300236408 ,Q3:2017.0 , IQR:7.0
42     1997
127    1975
255    1998
266    1997
275    1997
322    1998
325    1998
465    1975
472    1999
503    2069
575     207
603    1992
612    1999
661    1998
701    2060
711     202
757      13
805     207
825    1993
827    1999
Name: make_year, dtype: int64
Number of Outliers: 20


In [67]:
df.loc[503,'make_year']= 2012
df.loc[575,'make_year']= 2007
df.loc[701,'make_year']= 2003
df.loc[711,'make_year']= 2002
df.loc[757,'make_year']= 2013
df.loc[805,'make_year']= 2007
find_quartiles_and_outliers_print(['make_year'])



 For the make_year
Q1:2010.0 ,Q2:2013.3274231678488 ,Q3:2017.0 , IQR:7.0
42     1997
127    1975
255    1998
266    1997
275    1997
322    1998
325    1998
465    1975
472    1999
603    1992
612    1999
661    1998
825    1993
827    1999
Name: make_year, dtype: int64
Number of Outliers: 14


**Now for the mileage**

In [68]:
print(df[df['mileage'] > 100]['mileage'].describe())

count       11.000000
mean     12688.181818
std      23924.354774
min        120.000000
25%        270.000000
50%        400.000000
75%       8512.500000
max      69000.000000
Name: mileage, dtype: float64


Engine_cc Outliers 

In [69]:
def find_quartiles_and_outliers_print_range(columns,range):
    for column in columns:
        print(f"\n\n For the {column}")
        df[column] = pd.to_numeric(df[column],errors='coerce')
        Q1 = df[column].quantile(0.25)
        Q2 = df[column].mean()
        Q3 = df[column].quantile(0.75)
        IQR = Q3 - Q1
        print(f"Q1:{Q1} ,Q2:{Q2} ,Q3:{Q3} , IQR:{IQR}") 
        outliers = df[(df[column] < (Q1 - range* IQR)) | (df[column] > (Q3 + range * IQR))]
        print(outliers[column])
        print(f"Number of Outliers: {len(outliers)}")



In [70]:
find_quartiles_and_outliers_print_range(['engine_cc'],3)



 For the engine_cc
Q1:1197.0 ,Q2:1520.64689608637 ,Q3:1600.0 , IQR:403.0
2       2956.0
5       2956.0
56      3700.0
58      3783.0
88      2882.0
96      2900.0
113     3500.0
158    15000.0
181     3000.0
205     2967.0
228     4898.0
235     2900.0
244     3000.0
267     2900.0
279     5223.0
293     3000.0
311     2900.0
318     3000.0
327     2950.0
328     2999.0
335     3200.0
393     3200.0
439     3700.0
462     3200.0
496     5883.0
504     3200.0
519     4461.0
520     3200.0
535    10000.0
547     2956.0
566     3200.0
586     2956.0
631     2956.0
656     3000.0
704     3000.0
712     4000.0
721     3000.0
785     3000.0
Name: engine_cc, dtype: float64
Number of Outliers: 38


Clipping the Engine_cc at 6000

In [71]:
upper_bound=6000
df['engine_cc'] = df['engine_cc'].clip(upper=upper_bound)

**Keeping price,engine_cc outliers because they are genuine data, Capping the mileage outliers at the IQR**

In [72]:
Q1_m = df['mileage'].quantile(0.25)
Q3_m = df['mileage'].quantile(0.75)
IQR_m = Q3_m - Q1_m
df['mileage'] = df['mileage'].clip(
    lower=Q1_m - 1.5*IQR_m,
    upper=Q3_m + 1.5*IQR_m
)

**Handling the missing Values**

In [73]:
print(df['kilometer_run'].isna().sum())
print(df['price'].isna().sum())
print(df['engine_cc'].isna().sum())
print(df['mileage'].isna().sum())
print(df['make_year'].isna().sum())

0
6
105
110
0


Using median and mode for the numerical data. 

In [74]:
df['price'] = df['price'].fillna(df['price'].median())
df['engine_cc'] = df['engine_cc'].fillna(df['engine_cc'].mode()[0])
df['mileage'] = df['mileage'].fillna(df['mileage'].median())

Binary Encoding for Transmission Type 

In [75]:
print(df['transmisson_type'].value_counts())

transmisson_type
Manual           621
Auto             180
Power             28
PowerSteering     17
Name: count, dtype: int64


Keeping power steering also as manual because it is too small dataset to be another feature and mostly power steering is used in heavy vehicle. 

In [76]:
df['is_automatic'] = (df['transmisson_type'] == 'Auto').astype(int)
df.drop('transmisson_type', axis=1, inplace=True)

In [77]:
df.head()

,name,price,make_year,mileage,engine_cc,fuel,kilometer_run,drive,is_automatic
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,2475000.0,2015,14.0,1400.0,Petrol,42000.0,2WD,0
1,Excellent car on sale (Hyundai),750000.0,2005,11.0,1399.0,Petrol,87412.0,2WD,1
2,TATA 407 Container (Tata),700000.0,2013,12.0,2956.0,Diesel,60000.0,2WD,0
3,4x4 swaraj Mazda (Mahindra),600000.0,2017,14.0,1200.0,Diesel,33516.0,4WD,0
4,i20 Active good for used few time (Hyundai),375.0,2019,17.0,1200.0,Petrol,2400.0,2WD,1


Binary Encoding for Drive

In [78]:
print(df['drive'].value_counts())

drive
2WD          645
4WD          156
 Steering     28
              17
Name: count, dtype: int64


Keeping others at 2WD

In [79]:
df['is_4WD'] = (df['drive'] == '4WD').astype(int)
df.drop('drive', axis=1, inplace=True)

In [80]:
df.head()

,name,price,make_year,mileage,engine_cc,fuel,kilometer_run,is_automatic,is_4WD
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,2475000.0,2015,14.0,1400.0,Petrol,42000.0,0,0
1,Excellent car on sale (Hyundai),750000.0,2005,11.0,1399.0,Petrol,87412.0,1,0
2,TATA 407 Container (Tata),700000.0,2013,12.0,2956.0,Diesel,60000.0,0,0
3,4x4 swaraj Mazda (Mahindra),600000.0,2017,14.0,1200.0,Diesel,33516.0,0,1
4,i20 Active good for used few time (Hyundai),375.0,2019,17.0,1200.0,Petrol,2400.0,1,0


Encoding for Fuel 

In [81]:
print(df['fuel'].value_counts())

fuel
Petrol      602
Diesel      221
Electric     18
Hybrid        5
Name: count, dtype: int64


Ordinal Encoding

In [82]:
fuel_order = {
    "Petrol": 1,
    "Diesel": 2,
    "Electric": 3,
    "Hybrid": 4
}

df['fuel'] = df['fuel'].map(fuel_order)
df.head()

,name,price,make_year,mileage,engine_cc,fuel,kilometer_run,is_automatic,is_4WD
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,2475000.0,2015,14.0,1400.0,1,42000.0,0,0
1,Excellent car on sale (Hyundai),750000.0,2005,11.0,1399.0,1,87412.0,1,0
2,TATA 407 Container (Tata),700000.0,2013,12.0,2956.0,2,60000.0,0,0
3,4x4 swaraj Mazda (Mahindra),600000.0,2017,14.0,1200.0,2,33516.0,0,1
4,i20 Active good for used few time (Hyundai),375.0,2019,17.0,1200.0,1,2400.0,1,0


Finally changing make year into age. 

In [83]:
df['age'] = 2025 - df['make_year']
df.drop('make_year',axis=1,inplace=True)
df.head()

,name,price,mileage,engine_cc,fuel,kilometer_run,is_automatic,is_4WD,age
0,Hyundai | i20 Active S | TDi | 2015 | Hatchbac...,2475000.0,14.0,1400.0,1,42000.0,0,0,10
1,Excellent car on sale (Hyundai),750000.0,11.0,1399.0,1,87412.0,1,0,20
2,TATA 407 Container (Tata),700000.0,12.0,2956.0,2,60000.0,0,0,12
3,4x4 swaraj Mazda (Mahindra),600000.0,14.0,1200.0,2,33516.0,0,1,8
4,i20 Active good for used few time (Hyundai),375.0,17.0,1200.0,1,2400.0,1,0,6


**Saving the Cleaned Data**

In [84]:
df.to_csv("cleaned_data.csv", index=False)

In [85]:
print(len(df))

846
